src/channel-qvm.lisp

```diff
@@ -0,0 +1,336 @@
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple instr-qubits) params)
+    (when append-kraus-ops (apply-all-kraus-operators qvm instr append-kraus-ops))
+    (incf (pc qvm))
+    qvm))
+
+
+(defmethod transition :around ((qvm channel-qvm) (instr quil:measurement))
+  (let ((ret-qvm (call-next-method)))
+    (apply-classical-readout-noise ret-qvm instr)
+    ret-qvm))
+
+
+(defun matches-rule (rule instr position)
+  "check if rule is matched by instruction data and the position of the match request"
+  (let* ((noise-predicate (predicate (noise-predicate rule)))
+         (noise-position (noise-position (noise-predicate rule))))
+    (and (funcall noise-predicate instr) (string= position noise-position))))
+
+
+(defun match-rules (rules instr position)
+  "return the operation elements for the first matching rule"
+  (loop :for rule in rules
+        :when (matches-rule rule instr position)
+          :return (operation-elements rule)))
+
+
+(defgeneric apply-all-kraus-operators (qvm instr kraus-operators)
+  (:documentation "This function applies all the kraus operators in the list KRAUS-OPERATORS to the state of the system."))
+
+
+(defmethod apply-all-kraus-operators ((qvm channel-qvm) (instr quil:gate-application) kraus-operators)
+  "Apply all the kraus operators to the state of the system."
+  (loop :for kops :in kraus-operators
+        :do
+           (check-kraus-ops kops)
+           (apply-kraus-ops qvm instr kops)))
+
+
+(defgeneric apply-kraus-ops (qvm instr kraus-ops)
+  (:documentation "We randomly select a kraus operator from the list KRAUS-OPS to apply to the state of the program. We do this by transform sampling (cf [1]): We divide the unit interval [0,1] into n bins where the j-th bin size equals the probability p_j with which the j-th Kraus operator k_j should be applied. We know that the Kraus operators are normalized such that
+       p_j = <psi|k_j^H k_j |psi> 
+where x^H denotes hermitian conjugation of x and can therefore perform this sampling lazily: First generate a uniformly sampled random number r in [0,1]. Next, find j such that
+     sum_{k=1}^{j-1} p_k < r <= sum_{k=1}^{j} p_k
+This is possible by evaluating only all p_k for k<=j. Then pick this j as the choice of Kraus operator to apply.
+[1]: https://en.wikipedia.org/wiki/Inverse_transform_sampling)"))
+
+
+(defmethod apply-kraus-ops ((qvm channel-qvm) (instr quil:gate-application) kraus-ops)
+  (format t "applying ~a~%" instr)
+  (let* ((amps (%trial-amplitudes qvm))
+         (r (random 1.0d0))
+         (summed-p 0.0d0)
+         (logical-qubits (quil:application-arguments instr))
+         (qubits (mapcar #'quil:qubit-index logical-qubits)))
+    (check-type amps quantum-state)
+    (loop :for kj :in kraus-ops
+          :do
+             (replace amps (amplitudes qvm))
+             (apply-matrix-operator (magicl-matrix-to-quantum-operator kj) amps (apply #'nat-tuple qubits))
```

Comment:
This feels like an expensive loop. Can you just `incf summed-p`, then `:until .. :return kj`, then apply `kj`?

Reply:
I copied this function from the previous implementation in `noisy-qvm.lisp`. When reading the comments in that file (which I copied into this one), I think you need to apply `kj` in each iteration of the loop